src/main/io/gps.c

```diff
@@ -686,124 +1108,191 @@ void gpsInitUblox(void)
                         } else {
-                            gpsSetState(GPS_STATE_CONFIGURE);
+                            ubloxSetMessageRate(CLASS_NAV, MSG_NAV_DOP, 0);
                         }
+                    } else {
+                            ubloxSetMessageRate(CLASS_NAV, MSG_NAV_DOP, 1);
                     }
                     break;
+                case UBLOX_SAT_INFO:
+                    // enable by default, turned off when armed and receiving data to reduce in-flight traffic
+                    setSatInfoMessageRate(5);
+                    break;
+                case UBLOX_SET_NAV_RATE:
+                    // set the nav solution rate to the user's configured update rate
+                    gpsData.updateRate = gpsConfig()->gps_update_rate_hz;
+                    ubloxSetNavRate(gpsData.updateRate, 1, 1);
+                    break;
+                case UBLOX_MSG_CFG_GNSS:
+                    if ((gpsConfig()->sbasMode == SBAS_NONE) || (gpsConfig()->gps_ublox_use_galileo)) {
+                        ubloxSendPollMessage(MSG_CFG_GNSS);
+                    } else {
+                        gpsSetState(GPS_STATE_RECEIVING_DATA);
+                    }
+                    break;
+                    // TO DO: (separate PR) add steps that remove I2C or SPI data on ValSet aware units, eg
+                    // ubloxSetMessageRateValSet(CFG_MSGOUT_UBX_NAV_SAT_I2C, 0);
+                    // ubloxSetMessageRateValSet(CFG_MSGOUT_UBX_NAV_SAT_SPI, 0);
+                default:
+                    break;
             }
+        }
 
-            break;
+        switch (gpsData.ackState) {
+            case UBLOX_ACK_IDLE:
+                break;
+            case UBLOX_ACK_WAITING:
+                if (cmp32(millis(), gpsData.lastNavMessage) > UBLOX_ACK_TIMEOUT_MS){
+                    // give up, treat it like receiving ack
+                    gpsData.ackState = UBLOX_ACK_GOT_ACK;
+                }
+                break;
+            case UBLOX_ACK_GOT_ACK:
+                if (gpsData.state_position == UBLOX_MSG_CFG_GNSS) {
+                    // ublox should be initialised, try receiving
+                    gpsSetState(GPS_STATE_RECEIVING_DATA);
+                } else {
+                    // move forward one position, and clear the ack state
+                    gpsData.state_position++;
+                    gpsData.ackState = UBLOX_ACK_IDLE;
+                }
+                break;
+            case UBLOX_ACK_GOT_NACK:
+                // this is the tricky bit
+                // and we absolutely must get the unit type right
+                if (gpsData.state_position == UBLOX_DETECT_UNIT) {
+                    gpsSetState(GPS_STATE_CONFIGURE);
+                    gpsData.ackState = UBLOX_ACK_IDLE;
+                } else {
+                    // otherwise, for testing: just ignore nacks
+                    gpsData.state_position++;
```

Comment:
this may break badly for  gpsData.state_position == UBLOX_MSG_CFG_GNSS

Reply:
The whole ack/nack stuff needs attention.  It isn't working properly at present.  The timeout is incorrect, as a result we are not properly waiting for messages that we should be waiting for.  At present it 'works' but that's mostly luck, and because we ignore the result.  Maybe in a separate PR we can fix it properly.  Its on my todo list.

Reply:
See UBLOX_CONFIG_FINISHED ,